File: lib/state/app_state.dart
import 'package:flutter/foundation.dart';
import '../models/user_model.dart';
import '../models/notification_model.dart';

/// Central ViewModel for CareConnect (MVVM pattern).
///
/// Extends [ChangeNotifier] so that [GoRouter.refreshListenable] re-evaluates
/// redirect logic whenever authentication state changes.
class AppState extends ChangeNotifier {
  // ── Authentication ────────────────────────────────────────────────────────
  bool _isLoggedIn  = false;
  bool _isOnboarded = false;
  UserModel? _currentUser;

  bool get isLoggedIn  => _isLoggedIn;
  bool get isOnboarded => _isOnboarded;
  UserModel? get currentUser => _currentUser;

  /// Simulated login — accepts any non-empty email + password ≥ 6 chars.
  Future<bool> login(String email, String password) async {
    await Future.delayed(const Duration(milliseconds: 700));
    if (email.isNotEmpty && password.length >= 6) {
      _currentUser = UserModel(
        id: '1',
        name: 'Alex Johnson',
        email: email,
        joinDate: DateTime(2024, 1, 15),
        bloodGroup: 'B+',
        allergies: 'Penicillin, seafood',
      );
      _isLoggedIn = true;
      _loadSampleNotifications();
      notifyListeners();
      return true;
    }
    return false;
  }

  void logout() {
    _isLoggedIn  = false;
    _isOnboarded = false;
    _currentUser = null;
    _notifications.clear();
    notifyListeners();
  }

  // ── Onboarding ────────────────────────────────────────────────────────────
  void completeOnboarding({
    required bool leftHand,
    required bool highContrast,
    required bool largeTargets,
    required double textScale,
  }) {
    _leftHandMode    = leftHand;
    _highContrast    = highContrast;
    _largeTargets    = largeTargets;
    _textScaleFactor = textScale;
    _isOnboarded     = true;
    notifyListeners();
  }

  // ── Accessibility settings ─────────────────────────────────────────────────
  bool   _highContrast    = false;
  bool   _leftHandMode    = true;   // default: left-hand mode ON
  double _textScaleFactor = 1.0;
  bool   _largeTargets    = true;

  bool   get highContrast    => _highContrast;
  bool   get leftHandMode    => _leftHandMode;
  double get textScaleFactor => _textScaleFactor;
  bool   get largeTargets    => _largeTargets;

  void setHighContrast(bool v)      { _highContrast    = v; notifyListeners(); }
  void setLeftHandMode(bool v)      { _leftHandMode    = v; notifyListeners(); }
  void setTextScaleFactor(double v) { _textScaleFactor = v; notifyListeners(); }
  void setLargeTargets(bool v)      { _largeTargets    = v; notifyListeners(); }

  // ── Feature toggles ────────────────────────────────────────────────────────
  final Set<String> _enabledFeatures = {'left-nav', 'large-targets'};

  Set<String> get enabledFeatures    => Set.unmodifiable(_enabledFeatures);
  int  get enabledFeatureCount       => _enabledFeatures.length;
  bool isFeatureEnabled(String id)   => _enabledFeatures.contains(id);

  void toggleFeature(String id) {
    if (_enabledFeatures.contains(id)) {
      _enabledFeatures.remove(id);
    } else {
      _enabledFeatures.add(id);
    }
    notifyListeners();
  }

  // ── Profile ────────────────────────────────────────────────────────────────
  void updateProfile(String name, String email) {
    if (_currentUser != null) {
      _currentUser = _currentUser!.copyWith(name: name, email: email);
      notifyListeners();
    }
  }

  // ── Notifications ──────────────────────────────────────────────────────────
  final List<NotificationModel> _notifications = [];

  List<NotificationModel> get notifications =>
      List.unmodifiable(_notifications);

  int get unreadCount =>
      _notifications.where((n) => !n.isRead).length;

  void markRead(String id) {
    final idx = _notifications.indexWhere((n) => n.id == id);
    if (idx != -1) {
      _notifications[idx] = _notifications[idx].copyWith(isRead: true);
      notifyListeners();
    }
  }

  void markAllRead() {
    for (int i = 0; i < _notifications.length; i++) {
      _notifications[i] = _notifications[i].copyWith(isRead: true);
    }
    notifyListeners();
  }

  void _loadSampleNotifications() {
    _notifications.addAll([
      NotificationModel(
        id: '1',
        title: 'Accessibility Tip',
        body: 'Try enabling high-contrast mode for better visibility in bright environments.',
        timestamp: DateTime.now().subtract(const Duration(hours: 1)),
      ),
      NotificationModel(
        id: '2',
        title: 'New Feature Available',
        body: 'Voice Input Assistance is now available. Tap to learn more.',
        timestamp: DateTime.now().subtract(const Duration(hours: 3)),
      ),
      NotificationModel(
        id: '3',
        title: 'Weekly Reminder',
        body: 'Review your accessibility preferences to ensure they still meet your needs.',
        timestamp: DateTime.now().subtract(const Duration(days: 1)),
        isRead: true,
      ),
    ]);
  }
}

File: react-native/src/state/AppState.ts
export interface User {
  id: string;
  name: string;
  email: string;
  joinDate: Date;
  bloodGroup?: string;
  allergies?: string;
}

export interface Notification {
  id: string;
  title: string;
  body: string;
  timestamp: Date;
  isRead: boolean;
}

export interface AppStateType {
  isLoggedIn: boolean;
  isOnboarded: boolean;
  currentUser: User | null;
  highContrast: boolean;
  leftHandMode: boolean;
  textScaleFactor: number;
  largeTargets: boolean;
  enabledFeatures: Set<string>;
  notifications: Notification[];
}

const DEFAULT_STATE: AppStateType = {
  isLoggedIn: false,
  isOnboarded: false,
  currentUser: null,
  highContrast: false,
  leftHandMode: true,
  textScaleFactor: 1.0,
  largeTargets: true,
  enabledFeatures: new Set(['left-nav', 'large-targets']),
  notifications: [],
};

type Listener = () => void;

class AppStore {
  private state: AppStateType = {...DEFAULT_STATE, enabledFeatures: new Set(['left-nav', 'large-targets'])};
  private listeners: Set<Listener> = new Set();

  getState(): AppStateType {
    return this.state;
  }

  subscribe(listener: Listener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private notify() {
    this.listeners.forEach(l => l());
  }

  async login(email: string, password: string): Promise<boolean> {
    await new Promise(r => setTimeout(r, 700));
    if (email.trim().length > 0 && password.length >= 6) {
      this.state = {
        ...this.state,
        isLoggedIn: true,
        currentUser: {
          id: '1',
          name: 'Alex Johnson',
          email: email.trim(),
          joinDate: new Date(2024, 0, 15),
          bloodGroup: 'B+',
          allergies: 'Penicillin, seafood',
        },
        notifications: [
          {id: '1', title: 'Accessibility Tip', body: 'Try enabling high-contrast mode for better visibility.', timestamp: new Date(Date.now() - 3600000), isRead: false},
          {id: '2', title: 'New Feature Available', body: 'Voice Input Assistance is now available.', timestamp: new Date(Date.now() - 10800000), isRead: false},
          {id: '3', title: 'Weekly Reminder', body: 'Review your accessibility preferences.', timestamp: new Date(Date.now() - 86400000), isRead: true},
        ],
      };
      this.notify();
      return true;
    }
    return false;
  }

  logout() {
    this.state = {...DEFAULT_STATE, enabledFeatures: new Set(['left-nav', 'large-targets'])};
    this.notify();
  }

  completeOnboarding(prefs: {leftHand: boolean; highContrast: boolean; largeTargets: boolean; textScale: number}) {
    this.state = {
      ...this.state,
      isOnboarded: true,
      leftHandMode: prefs.leftHand,
      highContrast: prefs.highContrast,
      largeTargets: prefs.largeTargets,
      textScaleFactor: prefs.textScale,
    };
    this.notify();
  }

  setHighContrast(v: boolean) { this.state = {...this.state, highContrast: v}; this.notify(); }
  setLeftHandMode(v: boolean) { this.state = {...this.state, leftHandMode: v}; this.notify(); }
  setTextScaleFactor(v: number) { this.state = {...this.state, textScaleFactor: v}; this.notify(); }
  setLargeTargets(v: boolean) { this.state = {...this.state, largeTargets: v}; this.notify(); }

  toggleFeature(id: string) {
    const next = new Set(this.state.enabledFeatures);
    if (next.has(id)) { next.delete(id); } else { next.add(id); }
    this.state = {...this.state, enabledFeatures: next};
    this.notify();
  }

  isFeatureEnabled(id: string): boolean {
    return this.state.enabledFeatures.has(id);
  }

  markRead(id: string) {
    this.state = {
      ...this.state,
      notifications: this.state.notifications.map(n => n.id === id ? {...n, isRead: true} : n),
    };
    this.notify();
  }

  markAllRead() {
    this.state = {
      ...this.state,
      notifications: this.state.notifications.map(n => ({...n, isRead: true})),
    };
    this.notify();
  }

  get unreadCount(): number {
    return this.state.notifications.filter(n => !n.isRead).length;
  }

  updateProfile(name: string, email: string) {
    if (this.state.currentUser) {
      this.state = {...this.state, currentUser: {...this.state.currentUser, name, email}};
      this.notify();
    }
  }
}

export const appStore = new AppStore();
